src/jit/morph.cpp

```diff
@@ -11571,7 +11571,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             break;
 
         case GT_DIV:
-
+            /* Replace "val / dcon" with "val * (1.0 / dcon)" if dcon is a power of two */
+            if ((typ == TYP_FLOAT || typ == TYP_DOUBLE) && op2->IsCnsFltOrDbl())
+            {
+                double op2dVal = op2->gtDblCon.gtDconVal;
+                int64_t op2iVal = static_cast<int64_t>(op2dVal);
+                if (op2dVal == op2iVal && isPow2(op2iVal))
```

Comment:
I'd rather see this do an appropriate `isPow2(float)` and `isPow2(double)` check (which would also allow multiplication by reciprocal for values like `0.5`).

For normal finite numbers, this is basically just checking that the mantissa is zero. For subnormal finite numbers (where the exponent is zero), this is basically checking that the mantissa is itself a power of 2.

Reply:
Speaking of multiplication by the reciprocal of `0.5`. It looks like we don't do optimizations like converting `x * 2.0` to `x + x` or `x * 1.0` or `x * 0.0`

Reply:
Properly checking for power of 2 is basically a must. As is, the code triggers undefined behavior for large values and that's no good in a compiler.

Reply:
I am not sure I did it correctly could you please take a look at my isMod2 implementations for float/double? (can I use frexp here?)

Reply:
I suppose you could avoid bit twiddling and use the C math library functions instead. Something like:
```C++
bool HasInverse(double d)
{
    int exp;
    return (frexp(d, &exp) != 0) && (exp != 0) && isnormal(1.0 / d);
}
```
Though last time I checked VC++ had rather ugly implementations (non-intrinsics, inefficient) of the above functions.